norddrop/src/device/mod.rs

```diff
@@ -593,36 +594,79 @@ fn open_database(
     dbpath: &str,
     events: &EventDispatcher,
     logger: &slog::Logger,
+    moose: &Arc<dyn drop_analytics::Moose>,
 ) -> Result<drop_storage::Storage> {
     // Try opening the DB 3 times
     for i in 1..=3 {
         match drop_storage::Storage::new(logger.clone(), dbpath) {
             Ok(storage) => return Ok(storage),
-            Err(err) => warn!(logger, "Failed to open DB: {err}, already tried {i} times",),
+            Err(err) => error!(
+                logger,
+                "Failed to open DB at \"{dbpath}\": {err}, already tried {i} times",
+            ),
         }
     }
 
-    // Still problems? Let's try to delete the file
-    warn!(logger, "Removing old DB file");
-    if let Err(err) = std::fs::remove_file(dbpath) {
-        error!(
-            logger,
-            "Failed to opend DB and failed to remove it's file: {err}"
+    // If we can't even open the DB in memory, there is nothing else left to do,
+    // throw an error
+    if dbpath == ":memory:" {
+        let error = ffi::types::NORDDROP_RES_DB_ERROR;
+        let error_msg = "Failed to open in-memory DB";
+        moose.developer_exception(
+            -1,
+            error as i32,
+            "".to_string(),
+            error_msg.to_string(),
+            "Database Error".to_string(),
         );
-        return Err(ffi::types::NORDDROP_RES_DB_ERROR);
+
+        error!(logger, "{}", error_msg);
+        Err(error)
     } else {
-        // Inform app that we wiped the old DB file
-        events.dispatch(types::Event::RuntimeError {
-            status: drop_core::Status::DbLost,
-        });
-    };
+        moose.developer_exception(
+            -1,
+            ffi::types::NORDDROP_RES_DB_ERROR as i32,
+            "Initial DB open failed, recreating".to_string(),
+            "Failed to open database".to_string(),
+            "Database Error".to_string(),
+        );
+        // Still problems? Let's try to delete the file, provided it's not in memory
+        warn!(logger, "Removing old DB file");
+        if let Err(err) = std::fs::remove_file(dbpath) {
+            let error_msg = format!("Failed to open DB and failed to remove it's file: {err}");
+            moose.developer_exception(
+                -1,
+                ffi::types::NORDDROP_RES_DB_ERROR as i32,
```

Comment:
few very minor nitpicks:
- `-1` could be a constant with a constant why it's negative 1
- `ffi::types` could have a conversion to i32

Reply:
I could make a different function that uses the `arbitrary_value` and one that doesn't as well, and it's -1, because it's unused, the description of the function on our end is taken from the tracker itself

For the conversion, I thought it's very redundant to have a `Into<i32>` implementation, as the `ff::types` is `#[repr(C)]` with defined numeric values, including a conversion would actually be copy and pasting identical code and changing `as i32` to `.into()`

Reply:
Fair enough